### PR TITLE
soc: riscv: esp32c3: initial support for esp32c3 soc

### PR DESCRIPTION
**Pull request by uLipe**
Adds initial support for esp32c3 soc including its devkitm  board.

What is working:

* Basic kernel functionalities;
* Console log via printk;
* Generic samples such hello world and philosophers.

### CODE REVIEW

File: dts/riscv/espressif/esp32c3.dtsi [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	};
+
+	soc {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "simple-bus";
+		ranges;
+		
+		sram0: memory@3fc7c000 {
+			compatible = "espressif,esp32c3";
```
Comment (galak) [inline dts/riscv/espressif/esp32c3.dtsi:27]: This seems like a odd compatible.  Any reason this isn't `mmio-sram`?

File: arch/riscv/Kconfig [hunk 2]
```diff
@@ -90,6 +90,12 @@ config RISCV_SOC_INTERRUPT_INIT
 	  Enable SOC-based interrupt initialization
 	  (call soc_interrupt_init, within _IntLibInit when enabled)
 
+config RISCV_SOC_INIT_GP_VALUE
+	bool "Enable SOC-based global pointer register initialization"
+	help
+	  Enable SOC-based pointer register initialization
+	  (call __soc_get_gp_initial_value when initializing a thread)
+
```
Comment (katsuster) [inline arch/riscv/Kconfig:98]: I think this is not SoC support. It's better to separate another PR or commit.

File: arch/riscv/core/thread.c [hunk 3]
```diff
@@ -47,6 +47,10 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 	stack_init->a2 = (ulong_t)p2;
 	stack_init->a3 = (ulong_t)p3;
 
+#ifdef CONFIG_RISCV_SOC_INIT_GP_VALUE
+	stack_init->gp = __soc_get_gp_initial_value();
+#endif
+
```
Comment (katsuster) [inline arch/riscv/core/thread.c:53]: Same as "config RISCV_SOC_INIT_GP_VALUE".

File: soc/riscv/esp32c3/soc.c [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+		((rodata_reserved_start_align - SOC_DROM_LOW) / MMU_PAGE_SIZE) * sizeof(uint32_t);
+
+	esp32c3_rom_cache_set_idrom_mmu_size(cache_mmu_irom_size, CACHE_DROM_MMU_MAX_END - cache_mmu_irom_size);
+
+	/* set global esp32c3's INTC masking level */
+	esprv_intc_int_set_threshold(1);
+
+	/* Start Zephyr */
+	z_cstart();
+
```
Comment (katsuster) [inline soc/riscv/esp32c3/soc.c:101]: I guess this code want to avoid re-init C stack or bss areas in standard boot sequence (_PrepC() function).
However _PrepC does not only initialize the stack but also PMP, SoC depended initialization.

My concern is that can this code execute correctly even if changing implementation of _PrepC() in the future?

File: arch/riscv/Kconfig (hunk 2 above)
Comment (uLipe) [reply arch/riscv/Kconfig:98]: Thanks @katsuster I split this section in a separate commit.

File: arch/riscv/core/thread.c (hunk 3 above)
Comment (uLipe) [reply arch/riscv/core/thread.c:53]: This one, as well

File: dts/riscv/espressif/esp32c3.dtsi (hunk 1 above)
Comment (uLipe) [reply dts/riscv/espressif/esp32c3.dtsi]: Probably this was inherited when I imported some lines from esp32 dts, I corrected it. 

Thanks!

File: soc/riscv/esp32c3/soc.c [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		__asm__ __volatile__("la t0, _esp32c3_vector_table\n"
+							"csrw mtvec, t0\n");
+
+		/* Disable normal interrupts. */
+		csr_read_clear(mstatus, MSTATUS_MIE);
+
+		/* Zero out BSS.  Clobber _bss_start to avoid memset() elision. */
+		(void)memset(&_bss_start, 0, (&_bss_end - &_bss_start) * sizeof(_bss_start));
+		__asm__ __volatile__("" : : "g"(&_bss_start) : "memory");
+
```
Comment (katsuster) [inline soc/riscv/esp32c3/soc.c:58]: z_bss_zero() is not suitable?

File: soc/riscv/esp32c3/soc.c (hunk 4 above)
Comment (katsuster) [reply soc/riscv/esp32c3/soc.c:102]: Unfortunately, Xtensa (ESP32) and RISC-V is not same structure.
Xtensa does not have _PrepC() so no one can add codes. But RISC-V has _PrepC() and someone expect (Ex. PMP) to run initialization at there. If exp32c3 can guarantee to initialize all feature/modules at this _start function, it's no problem. But I think it is very uncomfortable for future esp32c3 maintainer...

What is current _PrepC() problem? If that is only BSS clear, shall we add new CONFIG for avoiding BSS clear?

File: soc/riscv/esp32c3/soc.c (hunk 5 above)
Comment (uLipe) [reply soc/riscv/esp32c3/soc.c:56]: Sure, this was also inherited from esp32 startup code, I added the z_bss_zero in place.

File: soc/riscv/esp32c3/soc.c (hunk 4 above)
Comment (uLipe) [reply soc/riscv/esp32c3/soc.c:96]: This make total sense, added the _PrepC at end of SoC specific startup code, delaying a bit the zeroing of the bss does not cause problem on the C3 special init code.

Thanks for the suggestion @katsuster

**Review by katsuster — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

**Comment by cwshu** [comment]
@uLipe
Based on my #36235 patch, I think we don't need to init GP register at thread init so that
the commit 'arch: riscv: ... gp register' is not needed. 
Could you also help to try #36235 patch on esp32c3? Any suggestion about this patch is welcome.

**Comment by uLipe** [comment]
@cwshu , If your patch gets merged in the upstream before this one, I surely can remove this commit, otherwise I prefer to keep the commit until this happens because without the gp properly set the esp32c3 will not work correctly since it uses the gp in a specific reserved area.

**Comment by cwshu** [comment]
@uLipe Thanks for your help. I could know that my patch about gp register also works on esp32c3 soc.

**Review by glaubermaroto — APPROVED** [review]

**Comment by uLipe** [comment]
Rebased to main branch, now this one could be merged because it will not break esp32 anymore.

cc: @sylvioalves  @galak  @nashif
